### PR TITLE
Integration with Apache Camel K, Debezium and Apache Kafka

### PR DESCRIPTION
**Pull request by machi1990**
DataStreaming solution using Debezium, AeroGear Graphback.
Implementation allows to stream data automatically from debezium without involving any stream processor. Graphback can utilize GraphQL for streaming and processing.

- [ ]  create sample Camel K integration to simulate "Entries" to complete test this. In the meantime I am using Mongo shell not so ideal. We can use GraphQL mutation till this is done.
- [x] Update OpenShift template to include Kafka , DBZ connector etc
- [ ] Polish the UI, maybe @kingsleyzissou or @wtrocki can step in here?

Attached are screenshots:

Screenshot from 2020-09-30 16-34-09
Screenshot from 2020-09-30 16-34-16
Screenshot from 2020-09-30 16-35-22

/cc @wtrocki

### CODE REVIEW

**Comment by machi1990** [comment]
1. The first image is a toast that appears on "all actions" page whenever there is a checkin
2. Once the toast is clicked, we land on the second page
3. still on the details page, which gets updated as the volunteer picks up actions. We can see this by the number of actions count which will increase as soon as the `VolunteerEntry` document is updated. Once the volunteer checks out, a toast appears indicating so. Clicking this will redirect to "all actions" page.

**Comment by kingsleyzissou** [comment]
@machi1990 happy to jump in and have a look at the UI

**Comment by machi1990** [comment]
> @machi1990 happy to jump in and have a look at the UI

Thanks @kingsleyzissou, really cool. I'll ping you once this branch is stable enough.

**Comment by wtrocki** [comment]
@machi1990 I will swap this to Ionic one when you happy.

**Comment by machi1990** [comment]
> @machi1990 I will swap this to Ionic one when you happy.

@wtrocki thanks, I still have https://github.com/aerogear/OpenVolunteerPlatform/blob/be0b7d97139b088f4b0b746ecf23b1eaccf2f19e/platform/server/README.md#todo to be wrapped.

**Comment by wtrocki** [comment]
@machi1990 Amazing work. I will be looking to push that on the openshift cluster we got. 
Keycloak and services are already there. We can set up things manually first and record some quick demo video early in the morning (I can do that no problem). 
Let's collaborate on this!

File: docs/full_stack_server.md [hunk 1]
```diff
@@ -98,7 +98,7 @@ services:
         ports:
             - 27017:27017
   keycloak:
-    image: jboss/keycloak:3.4.3.Final
+    image: quay.io/keycloak/keycloak:9.0.3
```
Comment (wtrocki) [inline docs/full_stack_server.md:101]: Yep.. Actually we working with keycloak 12 at the moment - they are really fast on releasing stuff.

File: platform/.openshift/keycloak.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
+---
+
+    apiVersion: template.openshift.io/v1
```
Comment (wtrocki) [inline platform/.openshift/keycloak.yml:3]: It will be more benefitial to use operator for us as we could create realm from Operator CR etc.

Comment (machi1990) [reply platform/.openshift/keycloak.yml:3]: Sure. 

I only needed this for my local testing. If we have keycloak, kafka services running then that's awesome.

You can still use the `yarn keycloak:init` command to initialise keycloak. All that's needed is the url and admin user credentials.

File: docs/full_stack_server.md (hunk 1 above)
Comment (machi1990) [reply docs/full_stack_server.md:101]: I had issues with latest release. I could not login after container was setup. Had to downgrade to 9.0.3. Could be my side or there is really an issue with the latest images.

**Comment by machi1990** [comment]
> @machi1990 Amazing work. I will be looking to push that on the openshift cluster we got.
> Keycloak and services are already there. We can set up things manually first and record some quick demo video early in the morning (I can do that no problem).
> Let's collaborate on this!

Only minor details (building new image, website and admin) are left. We can chat tomorrow / early monday about this.

File: platform/.openshift/keycloak.yml (hunk 2 above)
Comment (wtrocki) [reply platform/.openshift/keycloak.yml:3]: Yep. We can wrap that into cli and use some assumption `dataservices` namespace containing all URLs etc.

**Comment by machi1990** [comment]
@wtrocki @kingsleyzissou all good here and this branch is stable enough. I've pushed all the changes, plus image.

**Comment by wtrocki** [comment]
If we model this feature as addon we can easily merge that to master (while not requiring openshift to get things done)

**Comment by machi1990** [comment]
> If we model this feature as addon we can easily merge that to master (while not requiring openshift to get things done)

Yes, we might need feature flags / feature toggling for this.

**Comment by craicoverflow** [comment]
> The first image is a toast that appears on "all actions" page whenever there is a checkin

Struggled to see this toast appear for me. Looking at the code, a check-in is when on `CREATE_VOLUNTEERENTRY`? This is probably just me being unfamilar with this.

**Comment by wtrocki** [comment]
@craicoverflow This is cluster based effort and it involves some custom setup (Knative, Camel K). Things like RFID reader etc.
Going to share demo soon
Screenshot 2020-10-05 at 17 48 12

**Comment by machi1990** [comment]
> > The first image is a toast that appears on "all actions" page whenever there is a checkin
> 
> Struggled to see this toast appear for me. Looking at the code, a check-in is when on `CREATE_VOLUNTEERENTRY`? This is probably just me being unfamilar with this.

Is the Debezium connector setup as suggested here https://github.com/aerogear/OpenVolunteerPlatform/blob/a4c9492238888c73637429c879243ef9b2a5bc20/platform/server/README.md#debezium-connector-initialisation?  We can have quick chat tomorrow to have this properly setup on your "local" machine too and how you can simulate the checkin events via shell commands or api call.

**Comment by craicoverflow** [comment]
No, I missed that part. I will give it a go today!

**Comment by wtrocki** [comment]
Since this project is really overgrown what OVP is we have decided to move that to separate repository. 
OVP will still utilize Kafka streaming and Debezium but in form of extension that we going to build upon rather than thing that we have directly in the repository. There is also alarming number of supporting files that will not be needed once we demo it etc.

File: platform/server/src/KafkaSubEngine.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import { ObjectID } from 'mongodb';
+
+export interface IKafkaOptions {
+  topic: string
+  host: string
+  port: string
+  modelName: string
+}
+
+export class KafkaSubEngine extends PubSubEngine {
```
Comment (craicoverflow) [inline platform/server/src/KafkaSubEngine.ts:13]: What is the reason for using a custom engine over https://github.com/ancashoria/graphql-kafka-subscriptions?

The reason I am asking is I am looking to add a Kafka template to Graphback, do we need a custom one?

Comment (machi1990) [reply platform/server/src/KafkaSubEngine.ts:13]: @craicoverflow good question:

- This needed to be able to parse Debezium published change events into a structure that Graphback can understand. https://github.com/aerogear/OpenVolunteerPlatform/blob/a4c9492238888c73637429c879243ef9b2a5bc20/platform/server/src/KafkaSubEngine.ts#L105-L135

You can go with the library later on if changes will be handled by the application. But for this, the application has to handle topic creation too.

Comment (craicoverflow) [reply platform/server/src/KafkaSubEngine.ts:13]: Thanks for the info

**Pull request closed without merging**